### PR TITLE
Fix MPR7241: Pattern matching with mutable and lazy patterns is unsound

### PR DESCRIPTION
**Pull request by maranget**
Correct [PR#7241](https://caml.inria.fr/mantis/view.php?id=7241) --- side effect in guard segfaults because of PM compiler assuming that subject value does not change.

Correction consists in avoiding remembering what was matched. More precisely the compiler "forgets" what it learned during matching. This is performed by the various "rshift" functions.

Correction applies when
1. Some pattern has a mutable field.
2. And there is a guard or a lazy pattern.

Caveats
  PM may now fail (with Match_failure) although no warning has been given (better than segfault),
  see example in test suite.
  Suboptimal, one may think confining the memory erasure to mutable subterms.

### CODE REVIEW

**Comment by alainfrisch** [comment]
Can values be mutated without guards nor lazy patterns?  I was thinking about patterns that allocate (reading from float records or arrays => boxing) and could trigger finalizers or signal handlers, or yield control to a different thread.  Or are all allocations delayed until the guard/action?

**Comment by alainfrisch** [comment]
Also, what do you think about introducing a warning about patterns that deconstruct mutable fields and arrays?  This PR makes it less unsafe, but it's still not ideal to get a `Match_failure` and there does not seem to be any way to address that fully.

**Comment by maranget** [comment]
On your first comment I have no idea especially as regards allocating pattern matching (is there such a thing ?). As regards concurrent execution. if the thread that performs pattern matching can yield (can it ?), anything can happen.

On your second comment, I think shuch a warning would be obnoxious, being triggered much more often than useful.  However it would be very easy to implement...

**Comment by alainfrisch** [comment]
> (is there such a thing ?)

Yes, patterns that read floats from unboxed representations (float arrays, float records) allocate.  At least in native code, for capture variables which are known to be of type float (and which can potentially come from an unboxed representation), one could capture the unboxed version instead (i.e. simply read from unboxed representation, or unbox if the value comes from a boxed representation -- the two cases can be combined in a or-pattern).  This might be completely straightforward since unboxing is currently done much later than the compilation of pattern matching (and bytecode would not be supported).  But there will remain the case of arrays which can potentially hold floats (e.g. `function [|x|] -> Some x | _ -> None`).  Here, the pattern allocates only when the array hold floats, but the trick above would not work.

>  if the thread that performs pattern matching can yield (can it ?)

Yes, allocation can give control to another thread (or execute code through finalizers or signal handlers).

> anything can happen.

I think it would be safer to apply the same defensive mechanism that you suggest in this PR for any pattern that can possibly allocate, even with no guard or lazy pattern.

> I think shuch a warning would be obnoxious, being triggered much more often than useful. However it would be very easy to implement...

If the warning was restricted to cases that (i) read mutable fields and (ii) can execute code (lazy/guard/allocation), do you think it would be triggered too much?

**Comment by lpw25** [comment]
> I think it would be safer to apply the same defensive mechanism that you suggest in this PR for any pattern that can possibly allocate, even with no guard or lazy pattern.

I think that we should aim to treat these patterns correctly for the cases where we know unboxing is happening, as we can always inspect the float itself for the pattern and then only allocate when the pattern has been successfully matched.

As you say, this doesn't work for polymorphic patterns in arrays, so I guess being defensive is the only real option in that case. (Add that to the list of reasons why the float array hack should be removed).

**Comment by alainfrisch** [comment]
> then only allocate when the pattern has been successfully matched.

Note that to support or-pattern, we'd need to unbox boxed representations, as in:

``` ocaml
  function (Left [| x |] | Right x) -> x +. 1
```

and it will then be better to keep the unboxed binding (boxing only on demand).

I don't think this could be made to work easily for bytecode, though.

**Comment by lpw25** [comment]
> Note that to support or-pattern, we'd need to unbox boxed representations

I don't think OR-patterns make a difference, because you only need to extract the field if there is a pattern to match it against. So:

```
type t = { x : float }
let f = function { x } -> x
```

doesn't need to extract the field until after the match, whereas:

```
type t = { x : float }
let f = function
| { x = 0.0 } -> false
| _ -> true
```

does.

Actually, now that I think about it I suspect that float's in arrays are also fine. Since there are no polymorphic patterns which range over float and non-float, you can always delay extracting an item from a polymorphic array until after the match has completed.

**Comment by gasche** [comment]
I thought about this "only after matching" argument, but there is also the `when` guard, can't it result in matching more patterns in the future if it fails?

**Comment by lpw25** [comment]
> can't it result in matching more patterns in the future if it fails?

Yes, but `when` guards can allocate anyway so you don't care about allocating before them.

When I say "after matching" I mean the point after a pattern has been decided as matched, rather than the point after the (optional) `when` guard has been considered as well.

**Comment by alainfrisch** [comment]
@lpw25 How would you deal with:

``` ocaml
  function ((Left [| x |] | Right x),  (Left [| y ||] | Right y))-> x +. y
```

After detecting that the or-pattern succeeds, you need to remember which branches were taken (to know where to load x/y from and whether to allocate).  Or does the compilation of patterns "explodes" the four possible cases?

**Comment by lpw25** [comment]
> How would you deal with

Oh I see what you mean. Although that still seems resolvable. You could store an object and index for fetching the unboxed float, rather than a pointer to the boxed float.

**Comment by lpw25** [comment]
Except that only works if you _know_ about the boxing.

**Comment by lpw25** [comment]
So it seems that whilst or-patterns are fine and float arrays are fine, the combination of or-patterns and float arrays is problematic.

**Comment by gasche** [comment]
@alainfrisch I think that you would get four exits in this example (at least this is what compilation-by-matrices gives).

**Comment by alainfrisch** [comment]
> I think that you would get four exits in this example (at least this is what compilation-by-matrices gives).

Experiment shows it's not the case.  (The size of the lambda size grows linearly in the number of tuple components.)

>  the combination of or-patterns and float arrays is problematic.

Do you see any drawback in unboxing all capture variables of type float coming from at least one unboxed representation (in native code at least)?  So in:

``` ocaml
 function ((Left [| x |] | Right x),  (Left [| y |] | Right y))-> x +. y
```

one would unbox in the right branch of or-patterns.

This does not solve the problem of generic (float/non-float) arrays, but we know the real solution for that.

**Comment by maranget** [comment]
I would now tend to err on the safe side: cancel PM optimisation whenever a mutable field or an array is present in the patterns. It will then remain two issues
1. Confine the anti-optimisaton to subterm of the subject value that may change.
2. Warn users (should we?).

File: bytecomp/matching.ml [hunk 1]
```diff
@@ -2821,18 +2821,20 @@ and compile_no_test divide up_ctx repr mut partial ctx to_match =
 (* The entry points *)
 
 (*
-   If there is a guard in a matching or a lazy pattern,
-   then set exhaustiveness info to Partial.
+   If there is a mutable pattern set exhiastiveness to partial
```
Comment (alainfrisch) [inline bytecomp/matching.ml]: Typo.

**Comment by alainfrisch** [comment]
Ok, so the de-optimization is triggered by mutability even if no code can be executed given the current system (lazy/guards/allocations).  A typical example would be patterns such as `{desc = ..}` in the type-checker.  Do you believe this de-optimization could impact performance in a significant way?

**Comment by maranget** [comment]
I do not think performance will be affected significantly. If any performance problem, it can be alleviated by confining de-optimisation to subterms that can change, as I wrote above. In any case, correctness first.

**Comment by xavierleroy** [comment]
@maranget : as much as I trust your 6AM intuitions, I'd like to see some measurement of performance impact, e.g. on the benchmarks that were used for flambda.

**Comment by maranget** [comment]
Fair enough. Has anybody a pointer to those "benchmarks that were used for flambda". ?
CC @chambart

**Comment by chambart** [comment]
@maranget some of the informations are here. Sadly some benchmarks relying on core will have to be updated for the new version since the old ones can't compiler on 4.04.
https://www.typerex.org/operf-macro.html

**Comment by maranget** [comment]
Thanks,

--Luc

**Comment by maranget** [comment]
I have solved issue 1. above. That is "Confine the anti-optimisaton to subterm of the subject value that may change.". Unfortunately I could not install operf-macro (command "opam repo add operf-macro git:" does not work), so as to systematically test the performance degradation....
@chambart

File: bytecomp/matching.ml [hunk 2]
```diff
@@ -1647,8 +1668,8 @@ let make_record_matching loc all_labels def = function
           in
           let str =
             match lbl.lbl_mut with
-              Immutable -> Alias
-            | Mutable -> StrictOpt in
+              Immutable -> Alias,mut
+            | Mutable -> StrictOpt,true in
```
Comment (gasche) [inline bytecomp/matching.ml]: any idea why this variable is named "str"?

File: bytecomp/matching.ml [hunk 3]
```diff
@@ -2695,99 +2716,114 @@ let arg_to_var arg cls = match arg with
   The main compilation function.
    Input:
       repr=used for inserting debug events
+      mut= true <=> some field in pattern is mutable and there is are
+        guard or lazy patterns
```
Comment (gasche) [inline bytecomp/matching.ml]: Is the "and there are guard or lazy patterns" information really kept in the `mut` variable? In the code above at least it is only about mutability.

Also, the function `compile_match` below does not actually take `mut` argument. Does this comment (and "The 'mut' argument is here..." below) come from an older version of the patch?

File: bytecomp/matching.ml [hunk 4]
```diff
@@ -2836,51 +2874,47 @@ let find_in_pat pred =
...
 | Tpat_or _
 | Tpat_constant _ | Tpat_var _ | Tpat_any
   -> false
 
-let is_mutable p = find_in_pat have_mutable_field p
+let is_mutable p = find_in_pat is_mutable_pat p
 
 (* Downgrade Total when
    1. Matching accesses some mutable fields;
    2. And there are  guards or lazy patterns.
```
Comment (gasche) [inline bytecomp/matching.ml]: The new downgrading implementation does not take guards/lazy into account anymore, does it?

File: bytecomp/matching.ml (hunk 2 above)
Comment (maranget) [reply bytecomp/matching.ml]: No idea :)

File: bytecomp/matching.ml (hunk 4 above)
Comment (maranget) [inline bytecomp/matching.ml]: I have to clean comments up, thanks

But yes, the  new downgrading implementation does not take guards/lazy
into account. The reason being that I assume possible (if not likely...)
alteration of mutable subject values, due to signal handling,
finalizer calls, or any cause that may exist...

--Luc

**Comment by mshinwell** [comment]
I've added this PR to the benchmarking suite being used for Flambda.  A fair number of benchmarks aren't running yet since various packages need porting to trunk, but we may get some initial impression as to performance.

**Comment by mshinwell** [comment]
Hmm, it looks like there may be a performance problem with this one:

http://bench.flambda.ocamlpro.com/compare?test=2017-02-09-2300%2F4.05.0%2B%2Bpr717%2Bbench&reference=2017-02-09-2300%2F4.05.0%2Btrunk%2Bbench

**Comment by mshinwell** [comment]
@maranget Is there any hope of resolving this for 4.06?  (See benchmark link above)

File: Makefile.shared [hunk 5]
```diff
@@ -25,7 +25,7 @@ include stdlib/StdlibModules
 CAMLC=$(CAMLRUN) boot/ocamlc -g -nostdlib -I boot -use-prims byterun/primitives
 CAMLOPT=$(CAMLRUN) ./ocamlopt -g -nostdlib -I stdlib -I otherlibs/dynlink
 COMPFLAGS=-strict-sequence -principal -absname -w +a-4-9-41-42-44-45-48 \
-	  -warn-error A \
+	  -warn-error A-62 \
```
Comment (damiendoligez) [inline Makefile.shared]: This change is completely unneeded.

**Comment by gasche** [comment]
What is the status of this?

@maranget correct me if I'm wrong, but my understanding is that getting a fix that (1) is correct and (2) does not have a noticeable impact on performance (at least smaller than your proposed fix) is hard, possibly an open problem. In any case, that wouldn't be ready in time for 4.06.

For the release, we have to decide between using the current fix (despite its performance cost) or leaving the bug open. What is best?

**Comment by bluddy** [comment]
Would it be possible to optimize this a little by taking an immutable snapshot of the mutable fields as we enter a pattern match? Obviously this solution can't scale to huge mutable records, but for records with very few mutable fields, it should be fine.

We could improve this solution further by taking only the subset of the mutable fields that is examined by the pattern match (when we can determine said subset). This would be especially useful for array matching.

**Comment by maranget** [comment]
@gasche Performance impact of the fix is yet untested.
@bluddy Generally speaking, we are reluctant to introduce implicit allocation, as this would occur if we followed your idea of a snapshot. Besides, cost might be higher than the one of the current fix.

**Comment by bluddy** [comment]
@margaret since we know where the matching points are, we can preallocate this with the program. In many cases though it'll just use a few registers. For situations with few mutable elements and few observed elements, which I imagine to be the majority, this could well be faster. It gives us the atomicity we require.

**Comment by bluddy** [comment]
Thinking about it some more, snapshotting isn't a solution at all, since what the efficient pattern match does is essentially a form of snapshotting anyway -- we want to be consistent with the mutation that happens during a pattern match.

How about a conservative runtime solution? Have a global boolean flag indicating 'writing to a mutable field'. Pattern matching will clear the field, and check that it hasn't been set, since at no point during a pure pattern match should a write occur. If a write has taken place, the pattern match will revert to checking every field again. The flag should be residing in cache the whole time, and the writes to it can often be eliminated e.g. a function that knows it writes can just set it once upon entry; Flambda could eliminate most flag writes when inlining etc. For multicore, this flag would have to be thread-specific obviously.

There is also an additional heuristic for eliminating the flag update: if a function witnesses the creation of the record/array it's updating, it's not possible that it's being pattern-matched on right now, and it's safe to update without changing the flag.

**Comment by mshinwell** [comment]
@maranget I believe I did test the performance of this fix (see comment above from Feb 22).

**Comment by bluddy** [comment]
We shouldn't be surprised to see a serious performance impact. Programmers turn to mutability  often because they need extra performance in certain specific areas of the program. We're penalizing them now precisely in those places where they're 'telling us' they need performance. It's a targeted slowdown.

**Comment by maranget** [comment]
@mshinwell Sorry I have neglected the performance analysis. I see one way to enhance performance: restrict the fix to cases where is is assumed to be useful. That is activivate the de-optimization of matching on mutable fields when (1) there are guards (2) or lazy patterns (3) or explicit float pattern. Notice that, at the moment,  de-optimization is activated whenever there are patterns that match mutable data structures (ie record with mutable field or arrays).
I am  not so sure about point (3) above. Besides teh acceptance of pull request #1459 would probably mean a thorough rewriting of this pull request.

**Comment by bluddy** [comment]
I think restriction to guard patterns may already cause performance regressions.

BTW regarding my suggestion above, to keep track of mutation during pattern matching we could use a single bit of r15, the minor heap pointer on x86_64. It would mean that the flag will always be in the register set.

**Comment by bluddy** [comment]
Given the fact that we're headed towards multicore OCaml, and in a multithreaded model with sharing, this problem can be created by any thread mutating the values of another thread during a pattern match and there's nothing we can do about it short of completely destroying performance (ie. locking every matched object for the duration of the pattern match), I'm not sure we want to start damaging performance over this issue.

Instead, we could come up with incrementally better analysis runs with checking heuristics. If the user's match clause mutates an examined value, we could throw a compilation error. Of course, we'd be missing calls to functions from the match clauses, but more advanced iterations of this analysis could check or annotate functions with additional information about which arguments they modify. We could raise a warning if a conservative estimate shows a good chance of mutation happening over the matched value. If the user was 'smart' enough to save the matched-over value to a global and mutate it from there, well then that's almost like another thread swooping in and mutating the value during pattern match -- there's nothing we can do to help the user there. I think the same logic applies to GC finalizers -- they're so separate from the current context they may as well be in another thread (and in multicore, they likely _will_ be).

**Comment by bluddy** [comment]
Not sure what the status of this is, but trying to clarify this problem, there's one main situation where we have issues: when we match on a mutable value, and then later on dereference its pointer, either to extract a value or to check on an inner pattern. (This means, for example, that we don't care about mutable values that have no inner pointer) .For example, in

```
match !x, !y with
| None, Some a -> a (* 1 *)
| None, _ -> 0               (* 2 *)
| Some(Some b), _ -> b  (* 3 *)
```

!y is only examined and extracted from once. Once we've grabbed the pointer to `Some a`, we should have it and it won't go away even if the value of y is changed indirectly somehow. We could be wrong about the result, but we shouldn't be able to segfault. x is more problematic: since we already tested x for a `None`, we think we know it well enough to immediately dereference it at the `Some` branch later on. We could either re-examine !x (which I think is the current approach) or snapshot it beforehand. So the property we're looking for is a pattern that examines a mutable value more than once and dereferences its contained pointer at any point *after* the first examination. Unfortunately this property is recursive, applying as deeply as the pattern wishes to go, anywhere where a mutable value is found. (BTW, I apologize if this is obvious to many of you -- I'm thinking 'out loud' here).

So we have 2 approaches of dealing with this: pattern re-examination and snapshotting. By snapshotting, I'm referring mostly to pointers ie. in general we should have no extra allocations but we could have more spilling. It also means nothing more than keeping the pointer we used for our initial examination, or the unboxed representation of the float, around for longer. Clearly, in many cases, snapshotting will be faster. The downside of snapshotting is that we're keeping potentially old values, so we may be wrong -- but at least we won't segfault. The upside is that once we've factored in the cost of creating the snapshots, we can make as many assumptions as we wish about the values -- as many assumptions as the optimized pattern matching does today. Also, once we've passed the final dereference point of a particular value, we can discard its snapshot, further reducing the cost.

Either solution should be able to work at the same granularity i.e. only the subset of assumptions regarding patterns that match our criteria should be dropped with re-examination, and only the values that match our criteria should be snapshotted.

Additionally, as far as I can tell, with these approaches, we're making ourselves impervious to *any* problem, whether it's from a `when` clause, a GC finalizer or another thread in the multicore scenario.

**Comment by gasche** [comment]
The problem with snapshotting is that it sounds very hard to integrate in the pattern-matching algorithm. Naively one would think of binding each mutable dereference to a variable, and reusing that variable (instead of dereferencing again) on each further test on the same sub-part of the value. But there is no reason that those further tests would happen in the lexical scope of this variable. (Maybe one could, in theory, allocate those snapshot variables as mutable variables globally before generating any pattern-matching code, and assigning them on first read, with a dynamic test on later reads, but that would make each read more costly which is not nice.)

In general, the pattern-compilation algorithm is fairly sophisticated, so ideas that may appear to be simple can be arbitrarily difficult to implement. This is why Luc is generally working by trying the simplest possible ideas first, or rather the approaches that generate the least invasive patches -- and then growing in sophistication if required by performance results.

**Comment by bluddy** [comment]
@gasche I've quickly read through @maranget's 2001 paper, and I think the algorithm can be modified to snapshot. The classic algorithm would have been really easy to do this for, but even with the advanced algorithm (presented in the paper), where the problem is the monotonic gotos (aka numbered exceptions), I think you can find the lowest common ancestor of the reads and allocate an empty mutable variable there, then fill it in when reading the first time. Also, this only needs to be done for mutable values that hold pointers, thereby greatly reducing the impact.

As a side note, I know the paper is now 16 years old, but it would really have benefited from an example with the advanced algorithm. The classic algorithm is fairly intuitive and the examples given make it even clearer, but the advanced algorithm is never illustrated on an example, thus making it far harder to understand. There's also a question about how much this algorithm is of benefit on today's architectures, given that caching is far more dominant and jumping is potentially disruptive. For example, it was found on discuss.ocaml.org that the compilation of constant variants is quite a bit slower than polymorphic constant variants, at least on x86-64, since branch prediction is much faster on today's processors than indexing via a table. 

Finally, I think it's worth putting in the extra effort here, because this will come back to bite us with multicore. If we don't snapshot, users will experience segfaults with shared mutation in multicore OCaml, not to mention the rare cases mentioned above that cannot be detected with heuristics.

**Comment by maranget** [comment]
@shindere The CI failure on Travis probably is a consequence of a missing reference file for the flambda native compiler. Any clue?

**Comment by trefis** [comment]
I'll push a commit fixing the test failure with flambda.

@maranget the test failure aside, do you think the code is in a reasonable enough state that other people (e.g. I) can start reviewing it?

**Comment by trefis** [comment]
The test failure is due to the fact that the native code compiler with flambda enabled currently gives the warnings in the same order as in bytecode (I say currently because you aim to change that in #1649 ), and so in a different order than non-flambda.

We could probably add another reference file, or you could update your other PR to make the flambda codepath consistent with the other backends.

**Comment by maranget** [comment]
@trefis, yes please start review the codeif you wish.
@trefis again, after discussion with @shindere, I am  solving the missing reference file problem. You can refrain from solving it too...

**Comment by trefis** [comment]
Ack.

**Comment by trefis** [comment]
(Quick update: I'm finally starting to look at this)

File: bytecomp/matching.ml [hunk 6]
```diff
@@ -1638,20 +1658,20 @@ let make_record_matching loc all_labels def = function
           in
           let str =
             match lbl.lbl_mut with
-              Immutable -> Alias
-            | Mutable -> StrictOpt in
+              Immutable -> Alias,mut
+            | Mutable -> StrictOpt,no_opt && true in
```
Comment (trefis) [inline bytecomp/matching.ml]: The `&& true` part can go away.

File: bytecomp/matching.ml [hunk 7]
```diff
@@ -2827,87 +2859,136 @@ let find_in_pat pred =
...
-          pat_act_list
-      then Partial
-      else Total
+          (fun (pats,lam) -> may_call pats || is_guarded lam)
+          pat_act_list in
+      let partial = match partial with
+      | Partial -> Partial
+      | Total ->
+          if
+            mut || (match pat_act_list with [] -> true | _::_ -> false)
```
Comment (trefis) [inline bytecomp/matching.ml]: We've just inspected `pat_act_list`, so the line above can be simplified to `mut`.

File: bytecomp/matching.ml [hunk 8]
```diff
@@ -2920,16 +3001,16 @@ let partial_function loc () =
                Const_base(Const_int char)]))], loc)], loc)
 
 let for_function loc repr param pat_act_list partial =
-  compile_matching repr (partial_function loc) param pat_act_list partial
+  compile_matching loc repr (partial_function loc) param pat_act_list partial
 
 (* In the following two cases, exhaustiveness info is not available! *)
 let for_trywith param pat_act_list =
-  compile_matching None
+  compile_matching Location.none None
```
Comment (trefis) [inline bytecomp/matching.ml:3005]: I was a bit sad to see this `Location.none` considering we could, as other entry points do, take a loc parameter.
But this location is *currently* only used when emitting warning 63, which won't happen for `try ... with ...` as the matching is already considered partial anyway, so I guess it doesn't matter too much.

File: bytecomp/matching.ml [hunk 9]
```diff
@@ -2827,87 +2859,136 @@ let find_in_pat pred =
...
         match lbl.Types.lbl_mut with
         | Mutable -> true
         | Immutable -> false)
       lps
+| Tpat_array _ -> true
 | Tpat_alias _ | Tpat_variant _ | Tpat_lazy _
-| Tpat_tuple _|Tpat_construct _ | Tpat_array _
+| Tpat_tuple _|Tpat_construct _
 | Tpat_or _
 | Tpat_constant _ | Tpat_var _ | Tpat_any
```
Comment (trefis) [inline bytecomp/matching.ml:2874]: If I'm not mistaken, the above line isn't quite right in the absence of safe strings.

I suggest applying the following patch:
```diff
-| Tpat_constant _ | Tpat_var _ | Tpat_any
+| Tpat_var _ | Tpat_any
   -> false
+| Tpat_constant c -> begin                                                                                                             
+    match c with                                                                                                                       
+    | Const_string _ -> not Config.safe_string                                                                                         
+    | Const_int _ | Const_char _ | Const_float _                                                                                       
+    | Const_int32 _ | Const_int64 _ | Const_nativeint _ -> false                                                                       
+  end
```

**Review by trefis — COMMENTED** [review]
I left a few cosmetic comments in the code itself, here are some more general remarks:

- There is a decent amount of overlap between `Matching.may_call p && Matching.is_mutable p` and `not (Parmatch.inactive p)`, but they do not quite line up.
The version in `Matching` does seem correct for its use case, and `Parmatch.inactive` probably is too (though I haven't stopped to think about it); but it might make sense to share the overlapping parts instead of having them evolve separately.

- Again, there seem to be some overlap (not in terms of code this time, but in the intent) between `top_erase` and `set_args_erase_mutable`.
Is `set_args_erase_mutable` still needed? Wouldn't have `top_erase` already replaced "known" patterns (that can change) by `omega`s during the previous call to `combine`/`rshift`?

- On a related note, I sort of wonder if instead of forgetting some information (which might be/have been invalidated) when calling `combine`/`rshift` (i.e. "*context collection*" and "*context popping*" in the paper if I'm not mistaken), we could not push less information in the context. What would be the implications of such an approach?
(Note: I'm not suggesting doing that change, I'm just curious.)

- I am somewhat saddened by the fact that after `Parmatch` has concluded that a match is exhaustive, this information is not completely trusted. That has been the case for a long time, which was acceptable since it only impacted whether some optimizations would be performed. But now that it can emit a warning itself, we should probably consider improving parmatch to make it aware of these mutability concerns. For example, the following example that gets added to the testsuite by this PR currently raises the new warning 63:  
  ```ocaml
  type u = {a: bool; mutable b: int option}
  
  let f x =
    match x with
      {a=false} -> 0
    | {b=None} -> 1
    | _ when (x.b <- None; false) -> 2
    | {a=true; b=Some y} -> y
  ```
  And I find it somewhat sad that if I add a `| _ -> -1` case at the end of the matching, it will silence that warning, but be flagged as unused which it isn't. And I think we should be worried that we report useful clauses as unused.
(Note: if I put `.` instead of `-1` in there, then warning 63 gets emitted again, and I don't get a segfault when exercising that code path, just a Match_failure, which isn't *too* bad).

Apart from these remarks I think the code does what it claims: it disables optimizations which rely on informations that might change during matching.

Also, @maranget did some changes which should have a positive impact on performances since @mshinwell ran his benchmarks. I think it would be interesting to get new results after rebasing this PR on top of a recent trunk (which I did locally without getting any conflict).

File: bytecomp/matching.ml (hunk 9 above)
Comment (trefis) [reply bytecomp/matching.ml]: (The above should of course have read "the above line **isn't** quite right")

File: bytecomp/matching.ml (hunk 6 above)
Comment (maranget) [reply bytecomp/matching.ml]: Ok, wondering why such leftover code is still here.

File: bytecomp/matching.ml (hunk 7 above)
Comment (maranget) [reply bytecomp/matching.ml]: Ok, thanks "oeil d'aigle"

**Comment by trefis** [comment]
Btw, @mshinwell rerun the benchmark during the night, [the results](http://bench.flambda.ocamlpro.com/compare?test=2018-04-20-0924%2F4.08.0%2Bmutmatch%2Bbench&reference=2018-04-20-0924%2F4.08.0%2Btrunk%2Bbench#close) are better than previously (i.e. they are fine now).
Note: the "mutmatch" version is based on a rebased version of this PR + some trivial commits, available [here](https://github.com/trefis/ocaml/commits/pr7241).

File: bytecomp/matching.ml (hunk 8 above)
Comment (maranget) [reply bytecomp/matching.ml:3005]: I agree it does not matter, however I'll add a comment in matching.mli, so as to explain
the absence of this location parameter

File: bytecomp/matching.ml (hunk 9 above)
Comment (maranget) [reply bytecomp/matching.ml:2874]: Although you are right in principle, I'll leave the code as it is.
Problems (_i.e._ segfault) occur when the mutable sub-values change status, from scalar to pointer or the other way. In the case of a mutable string, characters may change to other characters, so nothing is to be feared.

**Comment by maranget** [comment]
I have integrated "trivial" cleanups (or argued why I did not integrate them).

I still have to look at the more general comments.

File: bytecomp/matching.ml (hunk 9 above)
Comment (trefis) [reply bytecomp/matching.ml:2874]: Yes, I wasn't worried about getting a segfault but about choosing "an incorrect branch" (which, ultimately, was the reason of the segfault in the initial report).

Comment (maranget) [reply bytecomp/matching.ml:2874]: Well, what the "correct" branch is in case the subject value is mutated is a moving
target. But well if insist, let us include your patch. Do you insist ?

Comment (trefis) [reply bytecomp/matching.ml:2874]: I agree, and no, I don't want to insist right now / on this particular point.

Comment (trefis) [reply bytecomp/matching.ml:2874]: (this particular point = mutable strings, in that context)

**Comment by trefis** [comment]
The embryo of a discussion we just had about choosing "the right branch" makes me wonder whether the following comment, which I made earlier, is still valid:

> Apart from these remarks I think the code does what it claims: it disables optimizations which rely on informations that might change during matching.

That's not quite right, in the sense that the compatibility check we use to decide whether we can reorder rows or not doesn't care about mutability. And one could argue that you should not allow reordering rows which look at the same mutable cell when mutation can happen.
To be clear: I don't think there is a soundness issue in allowing it (though I'll try and think about it a bit more), just that the behavior might surprise the user, and that we might want to change it at some point.

**Comment by trefis** [comment]
> And one could argue that you should not allow reordering rows which look at the same mutable cell when mutation can happen.

That of course seems impossible to track. Fortunately, I believe the condition should actually be "you never reorder rows looking at mutable cells when mutation can happen", which seems easier to enforce.

**Comment by xavierleroy** [comment]
What's the status on this PR?  Was the problem solved in one way or another?  Is the PR still relevant?

**Comment by trefis** [comment]
The problem is still present in trunk today. @gasche and I answered some of the questions that I asked during code review, while we were working on the pattern matching earlier this year.

That work is still ongoing, and we definitely hope to fix #7241 as a result of it. It's unclear to me whether our final fix will be a refresh of this PR, or something different, but what is clear is that this PR is unlikely to be merged.
I'll let @gasche close it if he agrees with this assessment.

**Comment by gasche** [comment]
I would vote for either "keep this one open as a reminder" or "close this one but open an issue". The first one doesn't require any work, and that is nice.

**Comment by gasche** [comment]
@maranget should we have a meeting to discuss the state of this PR? (cc maybe @trefis, @Octachron)

**Comment by gasche** [comment]
(A couple weeks ago I posted my current thoughts about this as a comment on the original issue: https://github.com/ocaml/ocaml/issues/7241#issuecomment-1594348867 )

**Comment by gasche** [comment]
We are not going to merge the present PR, but there is a lot of progress on other PRs to fix the same issue -- see #7241 for full details. Closing here.
